r3bbase/R3BTprevTnext.cxx

```diff
@@ -0,0 +1,170 @@
+{
+
+    FairRootManager* rootManager = FairRootManager::Instance();
+    if (rootManager == nullptr)
+    {
+        return kFATAL;
+    }
+
+    FairRuntimeDb* rtdb = FairRuntimeDb::instance();
+    if (rtdb == nullptr)
+    {
+        LOG(error) << "FairRuntimeDb not opened!";
+        return kFATAL;
+    }
+
+    fR3BEventHeader = dynamic_cast<R3BEventHeader*>(rootManager->GetObject("EventHeader."));
+    if (fR3BEventHeader == nullptr)
+    {
+        LOG(error) << "R3BTprevTnext::Init() EventHeader. not found";
+        return kFATAL;
+    }
+
+    fSamplerMapped = dynamic_cast<TClonesArray*>(rootManager->GetObject("SamplerMapped"));
+    if (fSamplerMapped == nullptr)
+    {
+        return kFATAL;
+    }
+
+    fSamplerMSMapped = dynamic_cast<TClonesArray*>(rootManager->GetObject("SamplerMSMapped"));
+    if (fSamplerMSMapped == nullptr)
+    {
+        return kFATAL;
+    }
+
+    fMSOffsetPar = dynamic_cast<R3BMSOffsetPar*>(rtdb->getContainer("MSOffsetPar"));
+    if (fMSOffsetPar == nullptr)
+    {
+        LOG(error) << "Could not find MSOffsetPar container!";
+        return kFATAL;
+    }
+
+    return kSUCCESS;
+}
+
+void R3BTprevTnext::Exec(Option_t* /*opt*/)
+{
+
+    fTprev = -1.0e3; // Initializaton of Tprev and Tnext
+    fTnext = -1.0e3; // If in the end you have this value is because this is an offspill or SAMP and/or SAMPMS are not
+                     // properly recorded
+    const Int_t sampHits = fSamplerMapped->GetEntriesFast();
+    const Int_t sampmsHits = fSamplerMSMapped->GetEntriesFast();
+    if ((sampHits < 1) || (sampmsHits != 1)) // If statement to check if the event is properly recorded
+    {
+        fR3BEventHeader->SetTprev(fTprev);
+        fR3BEventHeader->SetTnext(fTnext);
+        return;
+    }
+    const Double_t MSOffset = fMSOffsetPar->GetMSOffset();
+    Double_t SAMPTime = 0;
+    Double_t SAMPMSTime = 0;
+    R3BSamplerMappedData* SAMPMapped = nullptr;
+    R3BSamplerMappedData* SAMPMSMapped = nullptr;
+
+    SAMPMSMapped = dynamic_cast<R3BSamplerMappedData*>(fSamplerMSMapped->At(0));
+    SAMPMSTime = SAMPMSMapped->GetTime();
+
+    if (fR3BEventHeader->GetTpat() < fTpat && fR3BEventHeader->GetTrigger() == 1 && fR3BEventHeader->GetTpat() > 0)
+    {
+        if (sampHits == 1)
+        {                                      //
+            fTprev = 1e8;                      // If statement to check if your valid event
+            fTnext = 1e8;                      // have only 1 hit
+            fR3BEventHeader->SetTprev(fTprev); // If yes, Tprev and Tnext will have value 1e8
+            fR3BEventHeader->SetTnext(fTnext); //
+            return;
+        }
+        fTprev = 1e9; // Initialization with a different value after the first tests
+        fTnext = 1e9; // If you have this value in your tprev/tnext is because the previous or next hit is inside the
+                      // delta_clk range but is not a MS
+        for (Int_t i = 0; i < sampHits; ++i)
```

Comment:
Maybe put this for-loop into another member function and give it a nice name indicating the meaning?